# Small Size Image Classfication

In [1]:
from keras.preprocessing.image import ImageDataGenerator
from keras.models import Sequential
from keras.layers import Conv2D, MaxPooling2D
from keras.layers import Activation, Dropout, Flatten, Dense
from keras import backend as K

/Users/haoyushe/Code(Local)/tensorflow2/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.


In [3]:
# dimensions of our images.
img_width, img_height = 65, 65

train_data_dir = 'data/train'
validation_data_dir = 'data/validation'
# 90% train and 10% test
nb_train_samples = 480
nb_validation_samples = 120
epochs = 100
batch_size = 20

if K.image_data_format() == 'channels_first':
    input_shape = (3, img_width, img_height)
else:
    input_shape = (img_width, img_height, 3)

model = Sequential()
model.add(Conv2D(32, (3, 3), input_shape=input_shape))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))

model.add(Conv2D(32, (3, 3)))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))

model.add(Conv2D(64, (3, 3)))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))

model.add(Flatten())
model.add(Dense(64))
model.add(Activation('relu'))
model.add(Dropout(0.5))
model.add(Dense(2)) #6分类
model.add(Activation('softmax')) #采用Softmax

model.compile(loss='categorical_crossentropy', #多分类
              optimizer='rmsprop',
              metrics=['accuracy'])

# this is the augmentation configuration we will use for training
train_datagen = ImageDataGenerator(
    rescale=1. / 255,
    shear_range=0.2,
    zoom_range=0.2,
    horizontal_flip=True)

# this is the augmentation configuration we will use for testing:
# only rescaling
test_datagen = ImageDataGenerator(rescale=1. / 255)

train_generator = train_datagen.flow_from_directory(
    train_data_dir,
    target_size=(img_width, img_height),
    batch_size=batch_size,
    class_mode='categorical') #多分类

validation_generator = test_datagen.flow_from_directory(
    validation_data_dir,
    target_size=(img_width, img_height),
    batch_size=batch_size,
    class_mode='categorical') #多分类

model.fit_generator(
    train_generator,
    steps_per_epoch=nb_train_samples // batch_size,
    epochs=epochs,
    validation_data=validation_generator,
    validation_steps=nb_validation_samples // batch_size)


Found 480 images belonging to 2 classes.
Found 120 images belonging to 2 classes.
Epoch 1/100
24/24 [==============================] - 4s 153ms/step - loss: 0.7114 - acc: 0.5125 - val_loss: 0.6822 - val_acc: 0.9000
Epoch 2/100
24/24 [==============================] - 2s 93ms/step - loss: 0.6963 - acc: 0.5313 - val_loss: 0.6703 - val_acc: 0.9500
Epoch 3/100
24/24 [==============================] - 2s 98ms/step - loss: 0.7053 - acc: 0.6125 - val_loss: 0.6299 - val_acc: 0.8667
Epoch 4/100
24/24 [==============================] - 2s 97ms/step - loss: 0.6660 - acc: 0.6458 - val_loss: 0.6191 - val_acc: 0.8667
Epoch 5/100
24/24 [==============================] - 2s 92ms/step - loss: 0.6592 - acc: 0.6646 - val_loss: 0.6954 - val_acc: 0.5833
Epoch 6/100
24/24 [==============================] - 2s 93ms/step - loss: 0.5911 - acc: 0.6792 - val_loss: 0.1715 - val_acc: 0.9917
Epoch 7/100
24/24 [==============================] - 2s 95ms/step - loss: 0.6013 - acc: 0.6875 - val_loss: 0.4352 - val_acc: 

Epoch 62/100
24/24 [==============================] - 2s 98ms/step - loss: 0.0860 - acc: 0.9687 - val_loss: 0.0841 - val_acc: 0.9833
Epoch 63/100
24/24 [==============================] - 2s 99ms/step - loss: 0.1091 - acc: 0.9417 - val_loss: 0.1835 - val_acc: 0.9250
Epoch 64/100
24/24 [==============================] - 2s 98ms/step - loss: 0.0799 - acc: 0.9708 - val_loss: 0.0469 - val_acc: 0.9917
Epoch 65/100
24/24 [==============================] - 3s 110ms/step - loss: 0.1770 - acc: 0.9333 - val_loss: 0.3721 - val_acc: 0.9083
Epoch 66/100
24/24 [==============================] - 3s 119ms/step - loss: 0.1201 - acc: 0.9437 - val_loss: 0.0669 - val_acc: 0.9833
Epoch 67/100
24/24 [==============================] - 3s 117ms/step - loss: 0.0924 - acc: 0.9562 - val_loss: 0.0816 - val_acc: 0.9667
Epoch 68/100
24/24 [==============================] - 3s 107ms/step - loss: 0.0819 - acc: 0.9625 - val_loss: 0.1375 - val_acc: 0.9417
Epoch 69/100
24/24 [==============================] - 2s 98ms/ste

Found 480 images belonging to 2 classes.
Found 120 images belonging to 2 classes.
Epoch 1/100
24/24 [==============================] - 4s 153ms/step - loss: 0.7114 - acc: 0.5125 - val_loss: 0.6822 - val_acc: 0.9000
Epoch 2/100
24/24 [==============================] - 2s 93ms/step - loss: 0.6963 - acc: 0.5313 - val_loss: 0.6703 - val_acc: 0.9500
Epoch 3/100
24/24 [==============================] - 2s 98ms/step - loss: 0.7053 - acc: 0.6125 - val_loss: 0.6299 - val_acc: 0.8667
Epoch 4/100
24/24 [==============================] - 2s 97ms/step - loss: 0.6660 - acc: 0.6458 - val_loss: 0.6191 - val_acc: 0.8667
Epoch 5/100
24/24 [==============================] - 2s 92ms/step - loss: 0.6592 - acc: 0.6646 - val_loss: 0.6954 - val_acc: 0.5833
Epoch 6/100
24/24 [==============================] - 2s 93ms/step - loss: 0.5911 - acc: 0.6792 - val_loss: 0.1715 - val_acc: 0.9917
Epoch 7/100
24/24 [==============================] - 2s 95ms/step - loss: 0.6013 - acc: 0.6875 - val_loss: 0.4352 - val_acc: 0.9083
Epoch 8/100
24/24 [==============================] - 2s 95ms/step - loss: 0.5105 - acc: 0.7375 - val_loss: 0.2808 - val_acc: 0.9500
Epoch 9/100
24/24 [==============================] - 2s 96ms/step - loss: 0.5320 - acc: 0.7188 - val_loss: 0.4254 - val_acc: 0.9167
Epoch 10/100
24/24 [==============================] - 2s 96ms/step - loss: 0.4835 - acc: 0.7604 - val_loss: 0.1837 - val_acc: 1.0000
Epoch 11/100
24/24 [==============================] - 2s 96ms/step - loss: 0.4541 - acc: 0.7625 - val_loss: 0.2527 - val_acc: 0.8833
Epoch 12/100
24/24 [==============================] - 2s 97ms/step - loss: 0.4084 - acc: 0.8000 - val_loss: 0.4073 - val_acc: 0.8667
Epoch 13/100
24/24 [==============================] - 2s 99ms/step - loss: 0.4374 - acc: 0.7875 - val_loss: 0.3766 - val_acc: 0.9167
Epoch 14/100
24/24 [==============================] - 2s 97ms/step - loss: 0.3967 - acc: 0.8229 - val_loss: 0.1939 - val_acc: 0.9500
Epoch 15/100
24/24 [==============================] - 2s 99ms/step - loss: 0.4087 - acc: 0.8000 - val_loss: 0.2286 - val_acc: 0.8833
Epoch 16/100
24/24 [==============================] - 2s 100ms/step - loss: 0.3477 - acc: 0.8646 - val_loss: 0.4984 - val_acc: 0.8417
Epoch 17/100
24/24 [==============================] - 2s 98ms/step - loss: 0.3581 - acc: 0.8271 - val_loss: 0.1510 - val_acc: 0.9500
Epoch 18/100
24/24 [==============================] - 2s 97ms/step - loss: 0.3656 - acc: 0.8354 - val_loss: 0.1195 - val_acc: 0.9917
Epoch 19/100
24/24 [==============================] - 2s 98ms/step - loss: 0.3463 - acc: 0.8500 - val_loss: 0.4557 - val_acc: 0.8500
Epoch 20/100
24/24 [==============================] - 2s 98ms/step - loss: 0.3020 - acc: 0.8667 - val_loss: 0.2251 - val_acc: 0.9167
Epoch 21/100
24/24 [==============================] - 2s 99ms/step - loss: 0.3096 - acc: 0.8667 - val_loss: 0.1380 - val_acc: 0.9583
Epoch 22/100
24/24 [==============================] - 2s 98ms/step - loss: 0.2925 - acc: 0.8896 - val_loss: 0.3131 - val_acc: 0.8917
Epoch 23/100
24/24 [==============================] - 2s 100ms/step - loss: 0.2683 - acc: 0.8854 - val_loss: 0.1002 - val_acc: 0.9583
Epoch 24/100
24/24 [==============================] - 2s 98ms/step - loss: 0.2750 - acc: 0.8771 - val_loss: 0.1734 - val_acc: 0.9250
Epoch 25/100
24/24 [==============================] - 2s 98ms/step - loss: 0.2634 - acc: 0.8771 - val_loss: 0.1811 - val_acc: 0.9167
Epoch 26/100
24/24 [==============================] - 2s 101ms/step - loss: 0.2601 - acc: 0.8917 - val_loss: 0.2573 - val_acc: 0.8333
Epoch 27/100
24/24 [==============================] - 2s 100ms/step - loss: 0.2514 - acc: 0.8854 - val_loss: 0.0888 - val_acc: 0.9833
Epoch 28/100
24/24 [==============================] - 2s 98ms/step - loss: 0.2397 - acc: 0.9104 - val_loss: 0.3189 - val_acc: 0.8917
Epoch 29/100
24/24 [==============================] - 2s 99ms/step - loss: 0.2327 - acc: 0.8937 - val_loss: 0.1811 - val_acc: 0.9500
Epoch 30/100
24/24 [==============================] - 2s 101ms/step - loss: 0.1902 - acc: 0.9229 - val_loss: 0.1426 - val_acc: 0.9667
Epoch 31/100
24/24 [==============================] - 2s 98ms/step - loss: 0.2552 - acc: 0.9000 - val_loss: 0.0218 - val_acc: 0.9917
Epoch 32/100
24/24 [==============================] - 2s 99ms/step - loss: 0.2700 - acc: 0.8750 - val_loss: 0.2082 - val_acc: 0.9417
Epoch 33/100
24/24 [==============================] - 2s 99ms/step - loss: 0.1990 - acc: 0.9187 - val_loss: 0.2343 - val_acc: 0.8833
Epoch 34/100
24/24 [==============================] - 2s 98ms/step - loss: 0.2606 - acc: 0.9000 - val_loss: 0.1026 - val_acc: 0.9500
Epoch 35/100
24/24 [==============================] - 2s 100ms/step - loss: 0.1767 - acc: 0.9250 - val_loss: 5.1354 - val_acc: 0.5083
Epoch 36/100
24/24 [==============================] - 2s 101ms/step - loss: 0.3557 - acc: 0.8875 - val_loss: 0.0544 - val_acc: 0.9917
Epoch 37/100
24/24 [==============================] - 2s 97ms/step - loss: 0.1192 - acc: 0.9625 - val_loss: 0.1414 - val_acc: 0.9500
Epoch 38/100
24/24 [==============================] - 2s 99ms/step - loss: 0.1536 - acc: 0.9333 - val_loss: 0.0439 - val_acc: 0.9750
Epoch 39/100
24/24 [==============================] - 2s 98ms/step - loss: 0.2694 - acc: 0.9208 - val_loss: 0.0714 - val_acc: 0.9667
Epoch 40/100
24/24 [==============================] - 2s 99ms/step - loss: 0.1631 - acc: 0.9375 - val_loss: 0.0572 - val_acc: 0.9667
Epoch 41/100
24/24 [==============================] - 3s 113ms/step - loss: 0.1669 - acc: 0.9396 - val_loss: 0.0972 - val_acc: 0.9500
Epoch 42/100
24/24 [==============================] - 2s 100ms/step - loss: 0.1552 - acc: 0.9458 - val_loss: 0.0753 - val_acc: 0.9833
Epoch 43/100
24/24 [==============================] - 2s 96ms/step - loss: 0.1841 - acc: 0.9271 - val_loss: 0.1931 - val_acc: 0.9333
Epoch 44/100
24/24 [==============================] - 2s 100ms/step - loss: 0.1443 - acc: 0.9437 - val_loss: 0.0951 - val_acc: 0.9583
Epoch 45/100
24/24 [==============================] - 2s 98ms/step - loss: 0.2045 - acc: 0.9396 - val_loss: 0.2564 - val_acc: 0.9000
Epoch 46/100
24/24 [==============================] - 2s 98ms/step - loss: 0.1575 - acc: 0.9417 - val_loss: 0.1316 - val_acc: 0.9417
Epoch 47/100
24/24 [==============================] - 2s 99ms/step - loss: 0.1456 - acc: 0.9417 - val_loss: 0.0671 - val_acc: 0.9833
Epoch 48/100
24/24 [==============================] - 2s 98ms/step - loss: 0.1535 - acc: 0.9437 - val_loss: 0.2890 - val_acc: 0.9083
Epoch 49/100
24/24 [==============================] - 2s 97ms/step - loss: 0.1469 - acc: 0.9479 - val_loss: 0.1392 - val_acc: 0.9417
Epoch 50/100
24/24 [==============================] - 2s 97ms/step - loss: 0.2672 - acc: 0.9292 - val_loss: 0.2814 - val_acc: 0.9417
Epoch 51/100
24/24 [==============================] - 2s 97ms/step - loss: 0.1081 - acc: 0.9625 - val_loss: 0.2566 - val_acc: 0.9167
Epoch 52/100
24/24 [==============================] - 2s 98ms/step - loss: 0.1893 - acc: 0.9250 - val_loss: 0.0823 - val_acc: 0.9583
Epoch 53/100
24/24 [==============================] - 2s 98ms/step - loss: 0.1532 - acc: 0.9542 - val_loss: 0.1622 - val_acc: 0.9250
Epoch 54/100
24/24 [==============================] - 2s 97ms/step - loss: 0.1119 - acc: 0.9604 - val_loss: 0.0316 - val_acc: 0.9917
Epoch 55/100
24/24 [==============================] - 2s 99ms/step - loss: 0.1595 - acc: 0.9542 - val_loss: 0.5718 - val_acc: 0.9000
Epoch 56/100
24/24 [==============================] - 2s 99ms/step - loss: 0.1299 - acc: 0.9521 - val_loss: 0.0381 - val_acc: 0.9917
Epoch 57/100
24/24 [==============================] - 2s 99ms/step - loss: 0.1307 - acc: 0.9479 - val_loss: 0.1684 - val_acc: 0.9500
Epoch 58/100
24/24 [==============================] - 2s 98ms/step - loss: 0.1427 - acc: 0.9479 - val_loss: 0.1511 - val_acc: 0.9667
Epoch 59/100
24/24 [==============================] - 2s 97ms/step - loss: 0.1288 - acc: 0.9500 - val_loss: 0.3377 - val_acc: 0.9000
Epoch 60/100
24/24 [==============================] - 2s 100ms/step - loss: 0.1494 - acc: 0.9458 - val_loss: 0.0623 - val_acc: 0.9667
Epoch 61/100
24/24 [==============================] - 2s 99ms/step - loss: 0.1449 - acc: 0.9437 - val_loss: 0.8196 - val_acc: 0.8583
Epoch 62/100
24/24 [==============================] - 2s 98ms/step - loss: 0.0860 - acc: 0.9687 - val_loss: 0.0841 - val_acc: 0.9833
Epoch 63/100
24/24 [==============================] - 2s 99ms/step - loss: 0.1091 - acc: 0.9417 - val_loss: 0.1835 - val_acc: 0.9250
Epoch 64/100
24/24 [==============================] - 2s 98ms/step - loss: 0.0799 - acc: 0.9708 - val_loss: 0.0469 - val_acc: 0.9917
Epoch 65/100
24/24 [==============================] - 3s 110ms/step - loss: 0.1770 - acc: 0.9333 - val_loss: 0.3721 - val_acc: 0.9083
Epoch 66/100
24/24 [==============================] - 3s 119ms/step - loss: 0.1201 - acc: 0.9437 - val_loss: 0.0669 - val_acc: 0.9833
Epoch 67/100
24/24 [==============================] - 3s 117ms/step - loss: 0.0924 - acc: 0.9562 - val_loss: 0.0816 - val_acc: 0.9667
Epoch 68/100
24/24 [==============================] - 3s 107ms/step - loss: 0.0819 - acc: 0.9625 - val_loss: 0.1375 - val_acc: 0.9417
Epoch 69/100
24/24 [==============================] - 2s 98ms/step - loss: 0.1319 - acc: 0.9437 - val_loss: 0.7392 - val_acc: 0.8667
Epoch 70/100
24/24 [==============================] - 2s 97ms/step - loss: 0.0783 - acc: 0.9708 - val_loss: 0.0831 - val_acc: 0.9667
Epoch 71/100
24/24 [==============================] - 2s 97ms/step - loss: 0.1271 - acc: 0.9417 - val_loss: 0.0431 - val_acc: 0.9833
Epoch 72/100
24/24 [==============================] - 2s 97ms/step - loss: 0.0650 - acc: 0.9750 - val_loss: 0.0677 - val_acc: 0.9667
Epoch 73/100
24/24 [==============================] - 2s 97ms/step - loss: 0.1637 - acc: 0.9479 - val_loss: 0.0298 - val_acc: 0.9833
Epoch 74/100
24/24 [==============================] - 2s 96ms/step - loss: 0.0807 - acc: 0.9771 - val_loss: 0.1671 - val_acc: 0.9667
Epoch 75/100
24/24 [==============================] - 2s 97ms/step - loss: 0.1244 - acc: 0.9562 - val_loss: 0.5573 - val_acc: 0.9167
Epoch 76/100
24/24 [==============================] - 2s 97ms/step - loss: 0.0589 - acc: 0.9729 - val_loss: 0.1181 - val_acc: 0.9500
Epoch 77/100
24/24 [==============================] - 2s 97ms/step - loss: 0.0890 - acc: 0.9625 - val_loss: 0.2611 - val_acc: 0.9500
Epoch 78/100
24/24 [==============================] - 2s 96ms/step - loss: 0.0986 - acc: 0.9792 - val_loss: 0.2354 - val_acc: 0.9500
Epoch 79/100
24/24 [==============================] - 2s 99ms/step - loss: 0.1153 - acc: 0.9625 - val_loss: 0.0771 - val_acc: 0.9667
Epoch 80/100
24/24 [==============================] - 2s 99ms/step - loss: 0.0676 - acc: 0.9750 - val_loss: 0.0893 - val_acc: 0.9667
Epoch 81/100
24/24 [==============================] - 2s 99ms/step - loss: 0.1043 - acc: 0.9625 - val_loss: 0.1610 - val_acc: 0.9417
Epoch 82/100
24/24 [==============================] - 2s 98ms/step - loss: 0.1010 - acc: 0.9708 - val_loss: 0.4905 - val_acc: 0.9083
Epoch 83/100
24/24 [==============================] - 2s 98ms/step - loss: 0.0798 - acc: 0.9729 - val_loss: 0.0408 - val_acc: 0.9833
Epoch 84/100
24/24 [==============================] - 2s 98ms/step - loss: 0.0859 - acc: 0.9625 - val_loss: 0.0783 - val_acc: 0.9833
Epoch 85/100
24/24 [==============================] - 2s 98ms/step - loss: 0.0490 - acc: 0.9792 - val_loss: 0.7087 - val_acc: 0.9000
Epoch 86/100
24/24 [==============================] - 2s 96ms/step - loss: 0.0857 - acc: 0.9604 - val_loss: 0.2437 - val_acc: 0.9583
Epoch 87/100
24/24 [==============================] - 2s 100ms/step - loss: 0.0744 - acc: 0.9687 - val_loss: 0.8855 - val_acc: 0.8833
Epoch 88/100
24/24 [==============================] - 2s 97ms/step - loss: 0.0692 - acc: 0.9687 - val_loss: 1.5164 - val_acc: 0.8167
Epoch 89/100
24/24 [==============================] - 2s 100ms/step - loss: 0.0741 - acc: 0.9750 - val_loss: 0.0314 - val_acc: 0.9833
Epoch 90/100
24/24 [==============================] - 2s 98ms/step - loss: 0.0350 - acc: 0.9917 - val_loss: 0.0838 - val_acc: 0.9750
Epoch 91/100
24/24 [==============================] - 2s 99ms/step - loss: 0.1993 - acc: 0.9625 - val_loss: 0.7464 - val_acc: 0.8833
Epoch 92/100
24/24 [==============================] - 2s 100ms/step - loss: 0.1312 - acc: 0.9625 - val_loss: 1.0237 - val_acc: 0.8583
Epoch 93/100
24/24 [==============================] - 2s 99ms/step - loss: 0.0401 - acc: 0.9917 - val_loss: 0.6702 - val_acc: 0.9000
Epoch 94/100
24/24 [==============================] - 2s 99ms/step - loss: 0.0315 - acc: 0.9833 - val_loss: 0.3939 - val_acc: 0.9250
Epoch 95/100
24/24 [==============================] - 2s 98ms/step - loss: 0.1119 - acc: 0.9604 - val_loss: 0.2656 - val_acc: 0.9417
Epoch 96/100
24/24 [==============================] - 2s 98ms/step - loss: 0.0723 - acc: 0.9812 - val_loss: 0.7041 - val_acc: 0.9083
Epoch 97/100
24/24 [==============================] - 2s 98ms/step - loss: 0.0773 - acc: 0.9750 - val_loss: 0.3243 - val_acc: 0.9333
Epoch 98/100
24/24 [==============================] - 2s 98ms/step - loss: 0.0766 - acc: 0.9729 - val_loss: 0.0498 - val_acc: 0.9833
Epoch 99/100
24/24 [==============================] - 2s 98ms/step - loss: 0.0794 - acc: 0.9729 - val_loss: 0.0430 - val_acc: 0.9750
Epoch 100/100
24/24 [==============================] - 2s 99ms/step - loss: 0.0393 - acc: 0.9854 - val_loss: 1.5609 - val_acc: 0.8583